examples/gateway/proxy/routing.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/ipfs/go-ipns"
+	ipns_pb "github.com/ipfs/go-ipns/pb"
+	ic "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/libp2p/go-libp2p/core/peer"
+	"github.com/libp2p/go-libp2p/core/routing"
+)
+
+type proxyRouting struct {
+	host       string
+	httpClient *http.Client
+}
+
+func newProxyRouting(host string, client *http.Client) routing.ValueStore {
+	if client == nil {
+		client = http.DefaultClient
+	}
+
+	return &proxyRouting{
+		host:       host,
+		httpClient: client,
+	}
+}
+
+func (ps *proxyRouting) PutValue(context.Context, string, []byte, ...routing.Option) error {
+	return routing.ErrNotSupported
+}
+
+func (ps *proxyRouting) GetValue(ctx context.Context, k string, opts ...routing.Option) ([]byte, error) {
+	if !strings.HasPrefix(k, "/ipns/") {
+		return nil, routing.ErrNotSupported
+	}
+
+	k = strings.TrimPrefix(k, "/ipns/")
+	id, err := peer.IDFromBytes([]byte(k))
+	if err != nil {
+		return nil, err
+	}
+
+	return ps.fetch(ctx, id)
+}
+
+func (ps *proxyRouting) SearchValue(ctx context.Context, k string, opts ...routing.Option) (<-chan []byte, error) {
+	if !strings.HasPrefix(k, "/ipns/") {
+		return nil, routing.ErrNotSupported
+	}
+
+	k = strings.TrimPrefix(k, "/ipns/")
+	id, err := peer.IDFromBytes([]byte(k))
+	if err != nil {
+		return nil, err
+	}
+
+	ch := make(chan []byte)
+
+	go func() {
+		v, err := ps.fetch(ctx, id)
+		if err != nil {
+			close(ch)
+		} else {
+			ch <- v
+			close(ch)
+		}
+	}()
+
+	return ch, nil
+}
+
+func (ps *proxyRouting) fetch(ctx context.Context, id peer.ID) ([]byte, error) {
+	u, err := url.Parse(fmt.Sprintf("http://%s/ipns/%s", ps.host, peer.ToCid(id).String()))
+	if err != nil {
+		return nil, err
+	}
+	resp, err := ps.httpClient.Do(&http.Request{
+		Method: http.MethodGet,
+		URL:    u,
+		Header: http.Header{
+			"Accept": []string{"application/vnd.ipfs.ipns-record"},
+		},
+	})
+	if err != nil {
+		fmt.Println(err)
+		return nil, err
+	}
```

Comment:
Same here, need to check Status code of the response and return remote error if was anything other than 200.